src/drivers/intel/cavs/ipc.c

```diff
@@ -133,23 +156,44 @@ static void ipc_irq_handler(void *arg)
 	}
 }
 
-void ipc_platform_do_cmd(struct ipc *ipc)
+#if CAVS_VERSION > CAVS_VERSION_1_5
+static struct sof_ipc_cmd_hdr *ipc_cavs_read_msg(void)
 {
-	struct sof_ipc_reply reply;
-	int32_t err;
+	struct sof_ipc_cmd_hdr *hdr = _ipc->comp_data;
+	uint32_t dr;
+	uint32_t dd;
+	uint8_t *payload;
+
+	dr = ipc_read(IPC_DIPCTDR);
+	dd = ipc_read(IPC_DIPCTDD);
+
+	if (dr & IPC_DIPCTDR_MSG_REG) {
+		hdr->cmd = (dr & IPC_DIPCTDR_GLB_CMD_MASK) << 16;
+		hdr->size = sizeof(*hdr) + IPC_DIPCTDD_PAYLOAD_SIZE(dd);
+		payload = (uint8_t *)(hdr + 1);
+		payload[0] = (dd & 0x00FF0000) >> 16;
+		payload[1] = (dd & 0x0000FF00) >> 8;
+		payload[2] = (dd & 0x000000FF);
```

Comment:
Here, is the data layout defined by kernel ?  it could be playload = dd & 0xFFFFFF with refinement of ipc message layout in kernel

Reply:
No, it is not defined by the kernel. It was just my draft proposal, one of the ways to reconstruct the payload from the content transferred in the IPC register. Preserving the endianness as you suggested, might be better indeed, in case there is a 16-bit value transferred in the lower 16-bits rather then 3-bytes array as I had assumed.